README.md
# PhoneDTMF
**Arduino library to detect DTMF without any external devices.**
When you use a MCU to decode a DTMF, you don't need an external device like the MT8870. You just need the ADC input and you will be able to detect the tones with simple algorithms (see Goertzel Algorithm).  
This library is a simple but efficient way to detect telephone tones on microcontrollers like Arduino UNO, ESP8266, ESP32 or Teensy (Teensy has a more powerful library).  

<img alt="Display spectrum" src="circuits/oled.jpg" width="300" />

## Use library
~~~~cpp
#include <PhoneDTMF.h>
PhoneDTMF dtmf = PhoneDTMF();
void setup() {
    Serial.begin(9600);
    dtmf.begin(A0);
}
void loop() {
    uint8_t tones = dtmf.detect();
    char button = dtmf.tone2char(tones);
    if(button > 0) {
    	Serial.print(button); Serial.println(" pressed");
    }
    delay(1000);
}
~~~~

###### PhoneDTMF(int16_t sampleCount = 128, float amplify = 1.0f)
> Constructor to init the library.  
> Set ***`sampleCount`*** to define how many times the analog input should be sampled on each request. It must be at least 50 if you want a good result. 50 to 300 are standard values. Higher value gives you a better result but need more time to be executed. 128 samples with a frequence of 6000Hz need 128 \* 1 / 6000 = 21.3ms to be executed. So every call to `detect()` need 21ms.  
> Set ***`amplify`*** if your signal is not good or if your signal does not cover the entire amplitude. Standard values are from 1.0 to 4.0, if you need to amplify it more than 4 times, it would be better to put an opamp on your circuit.

###### uint16_t begin(uint8_t sensorPin, uint32_t maxFrequence = 0)
> begin the library  
> ***`sensorPin`*** is the analog pin used to detect the DTMF, must be an anlog pin.  
> ***`maxFrequence`*** is the frequence you want to use for detecting the signal. Leave 0 to use the maximal frequence available on your microcontroller or set the desired frequence (example 6000 for 6kHz).  
> Returns the real frequence (&lt;=MAX_FREQ) used to detect the signal.  

###### uint8_t detect(float* pMagnitudes = NULL, float magnitude = -1.0f)
> Call this function, every time you want measure the signal. As this is a sync-function, it need some times (~20ms) to return a value. After calling this function, you can use `getRealFrequence()` to see if it was done in the right time (just for a better result), this should be the same frequence as the one returned by the `begin()` function.  
> ***`pMagnitudes`*** is a pointer to an array to 8 floats. It returns the magnitudes of the 8 tones measured. Leave empty or NULL if you don't need the values.  
> ***`magnitude`*** is the minimum magnitude for detecting the tones. If a tone is detected, only 2 frequences are over this magnitude. With 12bits, 128 samples and 6000Hz this should be around 2000. If you are not sure, just leave -1 or empty, so it will try to set this value automatically.  
> Returns the tones detected as 8-bit number (example: 0010 0001)  
> **Remarks**: the tones detected must be exactly 2! If only 1 or 3 or 4 tones are detected, the detection failed or there wasn't a real DTMF tone to detect.  

###### char tone2char(uint8_t dtmf)
> If you want to decode the tones to the phone button, call this function.  
> ***`dtmf`*** is the value of the detected tones (see `detect()`)   
> Returns the character of your phone (example '1' or '#') or 0 if the tones are not valid.   


___
*There are also some utility functions, to get some debug informations:* 


**`uint32_t getSampleFrequence()`** returns the maximal sample frequence of your microcontroller (should be at least 6000Hz, if you want use this library)  
**`uint32_t getRealFrequence()`** returns the current frequence of sampling (should be less or equal to MAX_FREQ)  
**`uint16_t getAnalogCenter()`** returns the analog center of the measurements => when nothing is detected. This value is measured at beginning, when `begin()` is called  
**`uint16_t getBaseMagnitude()`** returns the magnitude when nothing is detected, detected when `begin()` is called  
**`uint16_t getMeasurementTime()`** returns the time, the `detect()` function will need to be executed. It is just a calculation (frequence * samples), it is not measured.  

### ESP32 tips:

**analog inputs**: Do not use ADC2 channels if you use WiFi, because they will be disabled. Use one of the ADC1 channels, like A0-A7:  
```cpp
#include <driver/adc.h>
void setup()
{
    adc1_config_width(ADC_WIDTH_BIT_12); // set 12 bit (0-4096)
    adc1_config_channel_atten(ADC1_CHANNEL_0, ADC_ATTEN_DB_0); // do not use attenuation
    dtmf.begin((uint8_t)ADC1_CHANNEL_0_GPIO_NUM); // Use ADC 1, Channel 0 (GPIO36 on Wroom32)
}
```
**generate DTMF**: ESP32 is enough powerful to generate also DTMF tones on the DAC output, but this will not be described here, just to know that the ESP can detect the DTMF tones but is also able to send them to the thelephone line.  


## Circuit
Just remember that you have to filter out the DC compoment and that the signal peaks should stay between 0V and VCC. You can break the ADC input of your microcontroller.  
So, to remove the DC voltage, add a capacitor (2.2uF) before you connect the line to your input.  
2 Zener  diodes and an opamp would are also good components to protect it. But if you are sure to have a right signal, you can measure directly on the bus.  
If you want to see some circuits, for example how to connect your Arduino directly to your analog telephone, check the <a href='circuits/'>**Circuits manual**</a>.  

<a href="circuits/"><img src='circuits/analogphone.png' width='200'></a>  


## Goertzel Algorithm
The Goertzel algorithm is a technique in digital signal processing (DSP) for efficient evaluation of the individual terms of the discrete Fourier transform (DFT).  
https://en.wikipedia.org/wiki/Goertzel_algorithm

Since the DTMF frequences are well known and are only 8 (for the phone), the algorithm can be shortened a little bit. This is the case in this library.  

## Links
Original code from this library: https://forum.arduino.cc/index.php?topic=121540.0  

DTMF on Wikipedia: https://en.wikipedia.org/wiki/Dual-tone_multi-frequency_signaling  

Nice documentation about DTMF: https://www.mdpi.com/2076-3417/9/3/422/pdf  

Frequence filtering: http://dataandsignal.com/frequency-detection/  


## Credits
This library is based on the code written by Jacob Rosenthal (June, 2012) that is based on the Kevin Banks code found at http://www.eetimes.com/design/embedded/4024443/The-Goertzel-Algorithm  

It was then modified by by El_Supremo and optimized for DTMF and Arduino UNO (December, 2016), see https://forum.arduino.cc/index.php?topic=121540.0  

Now, this library was modified to work with other processors too and with the possibility to detect the tones without find out the magnitude.  


CHANGELOG
phoneDTMF ?.?.? - ????.??.??

phoneDTMF 0.3.0 - 2019.09.03
* better magnitude calculation on low samples

phoneDTMF 0.2.0 - 2019.08.27
* removed 6000Hz as MAX_FREQ
* ADC center will be updated during measurement
* Debug mode will give you the result of the last 128 measurements, so you can choose between oscilloscope and spectrum on the OLED example

phoneDTMF 0.1.0 - 2019.08.21

* Initial release



src/PhoneDTMF.h
/*
Heavily modified by El_Supremo (Arduino forums) to decode DTMF tones
It is also public domain and provided on an AS-IS basis. There's no warranty
or guarantee of ANY kind whatsoever.

The Goertzel algorithm is long standing so see
http://en.wikipedia.org/wiki/Goertzel_algorithm for a full description.
It is often used in DTMF tone detection as an alternative to the Fast
Fourier Transform because it is quick with low overheard because it
is only searching for a single frequency rather than showing the
occurrence of all frequencies.
This work is entirely based on the Kevin Banks code found at
http://www.eetimes.com/design/embedded/4024443/The-Goertzel-Algorithm
so full credit to him for his generic implementation and breakdown. I've
simply massaged it into an Arduino library. I recommend reading his article
for a full description of whats going on behind the scenes.

Created by Jacob Rosenthal, June 20, 2012.
Released into the public domain.
*/

// ensure this library description is only included once
#ifndef DTMF_h
#define DTMF_h

// include types & constants of Wiring core API
#include "Arduino.h"

//#define DTMF_DEBUG

// library interface description
class PhoneDTMF
{
private:
  static const uint8_t  TONES = 8;
  const uint16_t DTMF_TONES[TONES] =
    {
      697,
      770,
      852,
      941,
     1209,
     1336,
     1477,
     1633
    };
  const uint8_t  DTMF_MAP[TONES * 2] = {0x11,0x21,0x41,0x12,0x22,0x42,0x14,0x24,0x44,0x28,0x81,0x82,0x84,0x88,0x18,0x48};
  const char  DTMF_CHAR[TONES * 2] =   { '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', 'A', 'B', 'C', 'D', '*', '#'};
    // Number of samples to detect tone (normally from 50 to 200)
  int16_t	_iSamplesCount;
	  // Maximal frequence for sampling => is calculated in the Init function
    // Note: this should be at least 6000Hz, since the measurement is done with this frequence
  uint32_t	_iSamplesFrequence;
	// The real sampling frequence from the last measurement
  uint32_t  _iRealFrequence;
	// If the real frequence is not like expected, a compensation delay is applied
  uint16_t	_iCompensation;
	// For detecting the tone, coefficient must be calculated => done in the Init function
  float		_afToneCoeff[TONES];
	// Q for each tone, to find out which frequence has detected. It will be reset in the resetDTMF function
  float		_afQ1[TONES];
	// Q for each tone, to find out which frequence has detected. It will be reset in the resetDTMF function
  float		_afQ2[TONES];
	// Pin used for the analog to digital conversion
  uint8_t	_Pin;
	// The value of ADC if nothing is detected (or in stand by) => this is calculated in the Init function
  int16_t	_iAdcCentre;
  float		_fAdcCentre;
	// Magnitude if nothing is detected (or in stand by) => calculated in the Init function
  float		_fBaseMagnitude;
    // Amplify measured signal, so it will be detected better (only if the signal does not cover the full amplitude)
  float     _fAmplifier;
#ifdef DTMF_DEBUG
  uint16_t  _aiAnalogData[128];
  uint8_t	_iDataIndex;
#endif
  
  // user-accessible "public" interface
public:
    
  PhoneDTMF(int16_t sampleCount = 128, float amplifier = 1.0f);
  uint16_t begin(uint8_t sensorPin, uint32_t maxFrequence = 0);
  uint8_t detect(float* pMagnitudes = NULL, float magnitude = -1.0f);
  char tone2char(uint8_t dtmf);
	
  uint32_t getSampleFrequence();
  uint32_t getRealFrequence();
  uint16_t getAnalogCenter();
  uint16_t getBaseMagnitude();
  uint16_t getMeasurementTime();

#ifdef DTMF_DEBUG
  uint16_t* getMeasurements();
#endif
  
  // library-accessible "private" interface
private:
  //El_Supremo change to int
  void ProcessSample(int16_t);
  void ResetDTMF(void);    
  void singleDetect();
  uint8_t calculateMeasurement(float* pRet = NULL, float magnitude = -1.0f);
};

#endif



src/PhoneDTMF.cpp
/*
Based on: https://github.com/jacobrosenthal/Goertzel
Heavily modified by Pete (El_Supremo on Arduino forums) to decode DTMF tones
It is also public domain and provided on an AS-IS basis. There's no warranty
or guarantee of ANY kind whatsoever.
This uses Digital Pin 4 to allow measurement of the sampling frequency

The Goertzel algorithm is long standing so see
http://en.wikipedia.org/wiki/Goertzel_algorithm for a full description.
It is often used in DTMF tone detection as an alternative to the Fast
Fourier Transform because it is quick with low overheard because it
is only searching for a single frequency rather than showing the
occurrence of all frequencies.
This work is entirely based on the Kevin Banks code found at
http://www.eetimes.com/design/embedded/4024443/The-Goertzel-Algorithm
so full credit to him for his generic implementation and breakdown. I've
simply massaged it into an Arduino library. I recommend reading his article
for a full description of whats going on behind the scenes.

Created by Jacob Rosenthal, June 20, 2012.
Released into the public domain.

*/

// include this library's description file
#include "PhoneDTMF.h"

  /// <summary>Constructor, you can set 2 params, if you don't like the standard behaviour</summary>
  /// <param name="sampleCount">number of measurements, default is 128 (with a frequence of 6000Hz it means 128 * 1/6000 = 21.3ms for each measurement). Normal values are from 50 to 200.</param>
  /// <param name="amplifier">if your signal is not good and you know that you just have 1/5 or 1/10 of the entire amplitude, you can amplify it via software with this parameter. Set a value from 1 to 5</param>
PhoneDTMF::PhoneDTMF(int16_t sampleCount, float amplifier)
{               
  _iSamplesCount = sampleCount; 
  _fAmplifier = amplifier;
  _iCompensation = 0;
  _fBaseMagnitude = 0.0f;
  _iSamplesFrequence = 6000;
  _iRealFrequence = 0;
  _fAdcCentre = 0.0f;
}

  /// <summary>Init the module (calculate frequence, ADC center, coefficients)</summary>
  /// <remarks>even if the frequence could be 25-50kHz, many microcontroller have a limited ADC reading frequence. 
  /// For example the ESP32 is able to execute readAnalogInput 25.000 times each second but internally it read the inputs only 6000 times.</remarks>
  /// <param name="sensorPin">the pin used to detect the DTMF</param>
  /// <param name="maxFrequence">maximal frequence to sample DTMF (set 0 if you want the maximal possible)</param>
uint16_t PhoneDTMF::begin(uint8_t sensorPin, uint32_t maxFrequence)
{
  _Pin = sensorPin;
  pinMode(_Pin, INPUT);
  delay(10);
  float magnitudes[TONES];
  uint32_t t1 = millis();
  for(uint16_t j=0;j<1000;j++)
  {
    singleDetect();
  }
  t1 = millis() - t1;
  float sampleFrequence = (1000.0f / t1) * 1000.0f;
  if (maxFrequence > 0 && sampleFrequence > maxFrequence) _iCompensation = ((1000.0f / (float)maxFrequence) - (1000.0f / (float)sampleFrequence)) * 1000.0f;
  else maxFrequence = sampleFrequence;
  
  float adcCenter = 0.0f;
  for (uint16_t j = 0; j < 100; j++)
  {
	  adcCenter += (float)analogRead(_Pin) / 100.0f;
  }
  _iAdcCentre = adcCenter;
  _iSamplesFrequence = (uint32_t)sampleFrequence;
  
  uint32_t oldFreq;
  do
  {
    oldFreq = _iRealFrequence;
    detect(magnitudes);
	if (_iRealFrequence > maxFrequence  && _iCompensation < 200)
	{
		_iCompensation++;
		continue;
	}
	if (_iRealFrequence < maxFrequence - 150 && _iCompensation > 0)
	{
		_iCompensation--;
		continue;
	}
  } while (oldFreq != _iRealFrequence);

  _fBaseMagnitude = 0.0f;
  for (uint8_t j = 0; j < TONES; j++)
	  _fBaseMagnitude += magnitudes[j];
  _fBaseMagnitude /= TONES;
  
  float	omega;
  // Calculate the coefficient for each DTMF tone
  for(uint8_t i=0; i<TONES; i++) 
  {
	  // other example: https://github.com/G6EJD/ESP32-Morse-Decoder/
	
    omega = (2.0f * PI * DTMF_TONES[i]) / _iRealFrequence;
    
    // DTMF detection doesn't need the phase.
    // Computation of the magnitudes (which DTMF does need) does not
    // require the value of the sin.
    // not needed    sine = sin(omega);
	  _afToneCoeff[i] = 2.0f * cos(omega);
  }
  ResetDTMF();

  return _iRealFrequence;
}

  /// <summary>Detect the frequences</summary>
  /// <param name="pMagnitudes">a pointer to an array of 8 floats. It will return the values of all frequences. NULL or leave empty if you don't need the values.</param>
  /// <param name="magnitude">set the magnitude to detect the tones. &lt;0 or leave empty if it should detect them automatically</param>
  /// <returns>tones detected as 8-bit number (1000 0010 means that the second and last frequence where detected)</returns>
uint8_t PhoneDTMF::detect(float* pMagnitudes, float magnitude)
{
  uint32_t fTemp = millis();
  for (uint16_t index=0; index< _iSamplesCount; index++)
  {
    singleDetect();
    delayMicroseconds(_iCompensation);
  }
  fTemp = millis() - fTemp;
  _iRealFrequence = 1.0f / ((fTemp / 1000.0f) / _iSamplesCount);
  _iAdcCentre = (_fAdcCentre + _iAdcCentre) / 2;
  return calculateMeasurement(pMagnitudes, magnitude);
}

  /// <summary>utility function to retrieve the pressed button with the current dtmf. See <see cref="detect(float* pMagnitudes, float magnitude)"/> to detect dtmf tones.</summary>
  /// <param name="dtmf">detected tones (8 bit flags, 0=not detected and 1=detected)</param>
  /// <returns>character of the button pressed (example '1' or '#')</returns>
char PhoneDTMF::tone2char(uint8_t dtmf)
{
  for(uint8_t j=0; j<TONES*2; j++) 
  {
    if(dtmf == DTMF_MAP[j]) return(DTMF_CHAR[j]);
  }
  return((char) 0);
}

  /// <summary>returns the maximal microcontroller sample frequence</summary>
uint32_t PhoneDTMF::getSampleFrequence()
{
	return _iSamplesFrequence;
}

  /// <summary>returns the analog middle when nothing is detected (used to detect tones)</summary>
uint16_t PhoneDTMF::getAnalogCenter()
{
	return _iAdcCentre;
}

  /// <summary>returns the real frequence by measuring the tones</summary>
uint32_t PhoneDTMF::getRealFrequence()
{
	return _iRealFrequence;
}

  /// <summary>returns magnitude when nothing is detected. Used to detect the tones automatically</summary>
uint16_t PhoneDTMF::getBaseMagnitude()
{
	return (uint16_t)_fBaseMagnitude;
}

  /// <summary>returns the time in ms for a single measurement</summary>
uint16_t PhoneDTMF::getMeasurementTime()
{
	return (uint16_t)((1000.0f / _iRealFrequence) * _iSamplesCount);
}

/*
  Private functions:
*/

uint8_t PhoneDTMF::calculateMeasurement(float* pRet, float magnitude)
{
  float dtmf_mag[TONES];
  float maxMag = _fBaseMagnitude * 3.0f;
  float midMag = 0.0f;
  for (uint8_t i = 0; i < TONES; i++)
  {
    dtmf_mag[i] = sqrt(_afQ1[i] * _afQ1[i] + _afQ2[i] * _afQ2[i] - _afToneCoeff[i] * _afQ1[i] * _afQ2[i]);
    midMag += dtmf_mag[i];
    if (maxMag < dtmf_mag[i]) maxMag = dtmf_mag[i];
  }
  if (pRet != NULL)
  {
    memcpy(pRet, dtmf_mag, sizeof(dtmf_mag));
  }
  ResetDTMF();

  uint8_t dtmf = 0;
  if (magnitude < 0.0f) 
  {
    magnitude = (midMag / TONES) * 2.5f;
  }

  for (uint8_t i = 0; i < TONES; i++)
  {
    if (dtmf_mag[i] > magnitude)
    {
      dtmf |= (0x01 << i);
    }
  }

  return dtmf;
}

void PhoneDTMF::singleDetect()
{
  ProcessSample(analogRead(_Pin));
}


/* Call this routine for every sample. */
//El_Supremo - change to int (WHY was it byte??)
  /// <summary>
void PhoneDTMF::ProcessSample(int16_t sample)
{
  float Q0;
  //EL_Supremo subtract adc_centre to offset the sample correctly
  for (uint8_t i = 0; i < TONES; i++)
  {
    Q0 = _afToneCoeff[i] * _afQ1[i] - _afQ2[i] + _fAmplifier * (sample - _iAdcCentre);
    _afQ2[i] = _afQ1[i];
    _afQ1[i] = Q0;
  }
  _fAdcCentre += (float)sample / (float)_iSamplesCount;
#ifdef DTMF_DEBUG
  if (_iDataIndex < 128)
  {
    _aiAnalogData[_iDataIndex++] = sample;
  }
#endif
}

/* Call this routine before every "block" (size=N) of samples. */
void PhoneDTMF::ResetDTMF(void)
{
  for (uint8_t i = 0; i < TONES; i++)
  {
    _afQ2[i] = 0;
    _afQ1[i] = 0;
  }
  _fAdcCentre = 0.0f;
#ifdef DTMF_DEBUG
  _iDataIndex = 0;
#endif
}

#ifdef DTMF_DEBUG
uint16_t* PhoneDTMF::getMeasurements()
{
  return _aiAnalogData;
}
#endif